### ISSUE TITLE
Update data structure of ccsd class

### ISSUE THREAD

**Issue by sunqm** [pyscf/pyscf#291]
Current ccsd and inherent CC classes use two attributes to hold the t1 and t2 amplitudes. It should be possible to compress the two amplitudes in one vector and, without loss of generality, to save only the vector in the class. When the individual amplitude is desired, a property method of the ccsd class can dispatch the amplitude from the compressed vector on the fly. The reasons I propose this change are:

1. A general CC class may be defined. For different CC flavors, like CCD, CCSD, CCSDT, ..., the class has the same structure and attributes. We don't have to add .t2, t3, t4... for each CC method. Same function signature for update_amps, energy, run_diis functions, etc. can be used while the body of these functions may be implemented differently in different classes.

2. The individual amplitudes have redundancy in the coefficients. If they were stored separately in the current implementation, a large amount of memory was wasted to hold the redundant coefficients. Compressing them to a vector can reduce roughly 30% runtime memory usage.

**Comment by sunqm** [comment]
I think this change should also make sense in the k-point version and finite temperature version. @jdmcclain47 @awhite862 you probably have a better opinion.

**Comment by tberkel** [comment]
I agree this makes a lot of sense.  It actually sounds very similar to how the R amplitudes are handled in EOM-CC.  In those cases, `r1` and `r2` are packed into a single vector that avoids the redundancy in the `r2` coefficients, using the `amplitudes_to_vector(r1,r2)` functions.

However, for point (2), concerning memory: for CCSD, are the t2 amplitudes (of size O^2V^2) really the memory bottleneck and not some of the ERIs?
